virtual/registry/sql_kv.go

```diff
@@ -0,0 +1,194 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to prepare range statement: %w", err)
+	}
+
+	return &sqlKV{
+		db: db,
+		sqlKVStatement: sqlKVStatement{
+			vst:        time.Now(),
+			tableName:  tableName,
+			setStmt:    setStmt,
+			getStmt:    getStmt,
+			deleteStmt: deleteStmt,
+			rangeStmt:  rangeStmt,
+		},
+	}, nil
+}
+
+func (sqlkv *sqlKV) beginTransaction(ctx context.Context) (transaction, error) {
+	tx, err := sqlkv.db.BeginTx(ctx, &sql.TxOptions{Isolation: sql.LevelSerializable})
+	if err != nil {
+		return nil, err
+	}
+
+	return &sqlTransactionKV{tx: tx, sqlKVStatement: &sqlKVStatement{
+		tableName:  sqlkv.tableName,
+		vst:        sqlkv.vst,
+		setStmt:    tx.StmtContext(ctx, sqlkv.setStmt),
+		getStmt:    tx.StmtContext(ctx, sqlkv.getStmt),
+		deleteStmt: tx.StmtContext(ctx, sqlkv.deleteStmt),
+		rangeStmt:  tx.StmtContext(ctx, sqlkv.rangeStmt),
+	}}, nil
+}
+
+func (sqlkv *sqlKV) transact(txfn func(transaction) (any, error)) (any, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+	tx, err := sqlkv.beginTransaction(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	rt, err := txfn(tx)
+	if err != nil {
+		return nil, err
+	}
+
+	err = tx.commit(context.TODO())
+	if err != nil {
+		return nil, err
+	}
+
+	return rt, nil
+}
+
+func (sqlkv *sqlKV) close(ctx context.Context) error {
+	return sqlkv.db.Close()
+}
+
+func (sqlkv *sqlKV) unsafeWipeAll() error {
+	_, err := sqlkv.db.Exec("DELETE FROM nola_kv;")
+	if err != nil {
+		return fmt.Errorf("failed to wipe table: %w", err)
+	}
+
+	return nil
+}
+
+type sqlTransactionKV struct {
+	tx *sql.Tx
+	*sqlKVStatement
+}
+
+func (sqltransactionkv *sqlKVStatement) put(ctx context.Context, key []byte, value []byte) error {
+	_, err := sqltransactionkv.setStmt.ExecContext(ctx, key, value, value)
+	return err
+}
+
+func (sqltransactionkv *sqlKVStatement) get(ctx context.Context, key []byte) ([]byte, bool, error) {
+	var data []byte
+
+	err := sqltransactionkv.getStmt.QueryRowContext(ctx, key).Scan(&data)
+	if errors.Is(err, sql.ErrNoRows) {
+		return nil, false, nil
+	} else if err != nil {
+		return nil, false, fmt.Errorf("failed to get kv: %w", err)
+	}
+
+	return data, true, nil
+}
+
+func (sqltransactionkv *sqlKVStatement) iterPrefix(ctx context.Context, prefix []byte, fn func(k, v []byte) error) error {
+	rows, err := sqltransactionkv.rangeStmt.QueryContext(ctx, append(prefix, '%'))
+	if err != nil {
+		return fmt.Errorf("failed to execute iterprefix: %w", err)
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var key []byte
+		var value []byte
+		if err := rows.Scan(&key, &value); err != nil {
+			return err
+		}
+
+		err = fn(key, value)
+		if err != nil {
+			return err
+		}
+	}
+
+	if err := rows.Err(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (sqltransactionkv *sqlTransactionKV) getVersionStamp() (int64, error) {
+	return time.Since(sqltransactionkv.vst).Microseconds(), nil
```

Comment:
I did a bit of googling and I think we can do something that is actually correct using Postgres using this function: https://pgpedia.info/p/pg_current_xact_id.html

https://www.postgresql.org/docs/current/functions-info.html#FUNCTIONS-PG-SNAPSHOT

double check me but it seems like calling that function in a new transaction is guaranteed to give us a monotonically increasing number like FDBs version stamp.

of course it won’t increase at the right “rate” which is a bit annoying…


At the very least instead of using local timestamp, can you run a transaction that calls a function in Postgres to get the time? This makes it so we only depend on a single nodes clock (Postgres) being consistent and increasing monotonically instead of every node in the clusters clock being accurate

Reply:
There is probably some smart way to combiners these two ideas but I’m not sure what it is. I’ll take calling the get time function in a Postgres transaction for now though

Reply:
> There is probably some smart way to combiners these two ideas but I’m not sure what it is. I’ll take calling the get time function in a Postgres transaction for now though

Yeah I'll do this for now.

From what I saw in a comment, this should be increasing at a rate of 1mil/sec, from what I understand FDB does dummy transactions under the hood which keeps this value increasing. In this case I'd need to see when PG updates this value if no changes are made.